wit/couchbase/subdocument.wit

```diff
@@ -0,0 +1,307 @@
+package wasmcloud:couchbase@0.1.0-draft;
+
+/// Perform sub-document lookup operations on Documents stored in a Couchbase cluster
+///
+/// Reference: https://docs.couchbase.com/go-sdk/current/howtos/subdocument-operations.html#retrieving
+interface subdocument-lookup {
+  use types.{
+    document-id, request-span, document, subdocument-path, document-error, retry-strategy, mutation-metadata
+  };
+
+  /// Options for performing a lookup
+  record lookup-get-options {
+    /// Whether the lookup refers to an extended attribute, rather than the document itself
+    xattr: option<bool>,
+  }
+
+  type lookup-exists-options = lookup-get-options;
+  type lookup-count-options = lookup-get-options;
+
+  /// Types of different lookup operations that can be specified
+  variant lookup-operation {
+    /// Whether the particular path exists
+    exists(tuple<subdocument-path, lookup-exists-options>),
+    /// Whether the particular path exists
+    get(tuple<subdocument-path, lookup-get-options>),
+    /// Whether the particular path exists
+    count(tuple<subdocument-path, lookup-count-options>),
+  }
+
+  /// Errors that occur during lookup operation
+  variant lookup-error {
+    /// A completely unexpected error
+    unexpected(string),
+  }
+
+  /// Options that control the lookup operation
+  record lookup-options {
+    /// Timeout that should be used, in nanoseconds
+    /// If not specified, the implementer *may* provide a default timeout.
+    timeout-ns: option<u64>,
+
+    /// How and whether to retry the operation
+    retry-strategy: option<retry-strategy>,
+
+    /// A known span to associate this lookup with
+    parent-span: option<request-span>,
+  }
+
+  /// Errors that occur during lookup for an individual subdocument path
+  variant subdocument-lookup-error {
+    /// The provided path does not exist in the document
+    path-does-not-exist(string),
+    /// Type Conflict between the path in the document and path command
+    path-mismatch(string),
+    /// A document error
+    document-error(document-error),
+    /// A completely unexpected error
+    unexpected(string),
+  }
+
+  /// Perform a lookup of one or more specs in a given document
+  lookup: func(
+    id: document-id,
+    operations: list<lookup-operation>,
+    options: lookup-options,
+  ) -> result<list<result<document, subdocument-lookup-error>>, lookup-error>;
+
+  /// Perform a lookup on a particular document from any replica server
+  lookup-in-any-replicas: func(
+    id: document-id,
+    operations: list<lookup-operation>,
+    options: lookup-options,
+  ) -> result<list<result<document, subdocument-lookup-error>>, lookup-error>;
```

Comment:
The return value is a [`LookupInReplicaResult`](https://pkg.go.dev/github.com/couchbase/gocb/v2#LookupInReplicaResult), which itself contains a list with the requested sub-doc values.

Reply:
Right, but given that any of the operations can fail independently, this is better represented as a list of the subdocuments and the errors that occurred doing the lookup on each one -- we can do better than the original SDK here, which would have left you with a `nil` in the list somewhere.

Reply:
Then how do you know at what index to look at to get the data you're interested in ?

Reply:
Sorry I'm not sure I understand -- we're talking about subdocument lookups right? Which Index is this referring to? Is that the collection?

Looking back at https://pkg.go.dev/github.com/couchbase/gocb/v2#LookupInReplicaResult, that's the single one right -- I *do* see that I'm missing a few fields there now, in particular `is-replica`, and `cas` along with the actual value... Is this what you mean? That `document` is missing both of those?

Also, as far as knowing which index to look at, the caller *should* know the list of operations that it wanted to perform, so if you're looking for the subdocument to a specific operation you can look at the list of results at the same index. WIT currently doesn't have a way to *ensure* these lists are the same size -- but given the structure they *should* be.

Reply:
I've fixed at least the missing information, `is-replica` and `cas` are now present in what's returned.

If we think the index issue can be resolved by using the index of the operation, that should do it?

Reply:
Hey @JesusTheHun any thoughts here? Is there any reason the server might return stuff in an order *not* specified by the caller? (that's something we can control but if it happens in the wild I'd love to know as well).

Reply:
The values are returned in order. I understood what you said as "if there is an error for a sub-op, we'll skip the value in the list".
The current signature looks good to me.

Reply:
Ahhhhh sorry yeah so how I see that in my head is that if there's an error, we'll return an error value for that entry (it'll take up the slot), so people can skip by it (though in programming languages without smooth destructuring it will be a little clunky)